src/Microsoft.AspNetCore.Blazor/Components/Provider.cs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.RenderTree;
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    /// <summary>
+    /// A component that provides one or more parameter values to all descendant components.
+    /// </summary>
+    public class Provider : IComponent
+    {
+        private RenderHandle _renderHandle;
+
+        /// <summary>
+        /// The content to which the value should be provided.
+        /// </summary>
+        [Parameter] private RenderFragment ChildContent { get; set; }
+
+        /// <summary>
+        /// The value to be provided.
+        /// </summary>
+        [Parameter] private object Value { get; set; }
+
+        /// <summary>
+        /// Optionally gives a name to the provided value. Descendant components
+        /// will be able to receive the value by specifying this name.
+        ///
+        /// If no name is specified, then descendant components will receive the
+        /// value based the type of value they are requesting.
+        /// </summary>
+        [Parameter] private string Name { get; set; }
+
+        /// <inheritdoc />
+        public void Init(RenderHandle renderHandle)
+        {
+            _renderHandle = renderHandle;
+        }
+
+        /// <inheritdoc />
+        public void SetParameters(ParameterCollection parameters)
+        {
+            // Implementing the parameter binding manually, instead of just calling
+            // parameters.AssignToProperties(this), is just a very slight perf optimization
+            // and makes it simpler impose rules about the params being required or not.
+
+            var hasSuppliedValue = false;
+            Value = null;
+            ChildContent = null;
+            Name = null;
+
+            foreach (var parameter in parameters)
```

Comment:
Unrelated node, I suspect `ParameterEnumerator` would be improved by switching to a pattern like this: https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs#L598 (Assign `Current` inside `MoveNext()`) - `Current` is not going to get inlined anyway with an exception inside it.

Reply:
Sounds good. Done.